File: KITE-Mediasoup-Test/src/main/java/io/cosmosoftware/kite/mediasoup/checks/AllVideoCheck.java
package io.cosmosoftware.kite.mediasoup.checks;

import io.cosmosoftware.kite.entities.Timeouts;
import io.cosmosoftware.kite.exception.KiteTestException;
import io.cosmosoftware.kite.interfaces.Runner;
import io.cosmosoftware.kite.mediasoup.pages.MediasoupPage;
import io.cosmosoftware.kite.report.Reporter;
import io.cosmosoftware.kite.report.Status;
import io.cosmosoftware.kite.steps.TestStep;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import java.util.List;

import static io.cosmosoftware.kite.util.TestUtils.videoCheck;
import static io.cosmosoftware.kite.util.TestUtils.waitAround;

public class AllVideoCheck extends TestStep {

  
  private final int numberOfParticipants;
  private final MediasoupPage mediasoupPage;

  public AllVideoCheck(Runner runner, int numberOfParticipants) {
    super(runner);
    this.numberOfParticipants = numberOfParticipants;
    this.mediasoupPage = new MediasoupPage(runner);
  }

  @Override
  public String stepDescription() {
    return "Check the other videos are being received OK";
  }

  @Override
  protected void step() throws KiteTestException {
    try {
      //wait a while to allow all videos to load.
      waitAround(numberOfParticipants * 3 * Timeouts.ONE_SECOND_INTERVAL);
      logger.info("Looking for video elements");
      List<WebElement> videos = mediasoupPage.getVideoElements();
      if (videos.size() < numberOfParticipants) {
        throw new KiteTestException(
            "Unable to find " + numberOfParticipants + " <video> element on the page. No video found = "
              + videos.size(), Status.FAILED);
      }
      String videoCheck = "";
      boolean error = false;
      for (int i = 1; i < numberOfParticipants; i++) {
        String v = videoCheck(webDriver, i);
        videoCheck += v;
        if (i < numberOfParticipants - 1) {
          videoCheck += "|";
        }
        if (!"video".equalsIgnoreCase(v)) {
          error = true;
        }
      }
      if (error) {
        reporter.textAttachment(report, "Reveived Videos", videoCheck, "plain");
        throw new KiteTestException("Some videos are still or blank: " + videoCheck, Status.FAILED);
      }
    } catch (KiteTestException e) {
      throw e;
    } catch (Exception e) {
      throw new KiteTestException("Error looking for the video", Status.BROKEN, e);
    }
  }
}

File: KITE-Hangouts-Test/js/checks/AllVideoCheck.js
const {TestStep, Status, KiteTestError} = require('kite-common');

const executeStep = async function(stepInfo) {
  //first video is the fullscreen
  //second video is "display": "none"
  //third video is "You" (the publisher)
  const noVideos = stepInfo.page.getVideos().length;
  const FIRST_VIDEO_INDEX = noVideos > 1 ? 2 : 0;
  let result = "";
  let tmp;
  let error = false;
  for(i = FIRST_VIDEO_INDEX + 1; i < noVideos; i++) {
    let tmp = await stepInfo.page.videoCheck(stepInfo, i);
    console.log('received video[' + i + '] ' + tmp);
    result += tmp;
    if (i < noVideos - 1) {
      result += ' | ';
    }
    if (tmp != 'video') {
      error = true;
    }
  }

  if (error) {
    stepInfo.testReporter.textAttachment(stepInfo.report, "Received videos", result, "plain");
    throw new KiteTestError(Status.FAILED, "Some videos are still or blank: " + result);
  }
}

class AllVideoCheck extends TestStep {
  constructor(kiteBaseTest) {
    super();
    this.driver = kiteBaseTest.driver;
    this.numberOfParticipant = kiteBaseTest.numberOfParticipant;
    this.timeout = kiteBaseTest.timeout;
    this.page = kiteBaseTest.page;
    this.usersPerRoom = kiteBaseTest.payload.usersPerRoom;

    // Test reporter if you want to add attachment(s)
    this.testReporter = kiteBaseTest.reporter;
  }

  stepDescription() {
    return "Check the other videos are being received OK";
  }

  async step() {
    try {
      await executeStep(this);

    } catch (error) {
      console.log(error);
      if (error instanceof KiteTestError) {
        throw error;
      } else {
        throw new KiteTestError(Status.BROKEN, "Error looking for the video");
      }
    }
  }
}

module.exports = AllVideoCheck;